components/service/glean/src/main/java/mozilla/components/service/glean/storages/GenericScalarStorageEngine.kt

```diff
@@ -0,0 +1,183 @@
+internal typealias GenericStorageMap<T> = MutableMap<String, GenericDataStorage<T>>
+
+abstract class GenericScalarStorageEngine<ScalarType> : StorageEngine {
+    override lateinit var applicationContext: Context
+
+    // Let derived class define a logger so that they can provide a proper name,
+    // useful when debugging weird behaviours.
+    abstract val logger: Logger
+
+    protected val userLifetimeStorage: SharedPreferences by lazy { deserializeUserLifetime() }
+
+    // Store a map for each lifetime as an array element:
+    // Array[Lifetime] = Map[StorageName, ScalarType].
+    protected val dataStores: Array<GenericStorageMap<ScalarType>> =
+        Array(Lifetime.values().size) { mutableMapOf<String, GenericDataStorage<ScalarType>>() }
+
+    /**
+     * Implementor's provided function to convert deserialized 'user' lifetime
+     * data to the destination ScalarType.
+     *
+     * @param value loaded from the storage as [Any]
+     *
+     * @return data as [ScalarType] or null if deserialization failed
+     */
+    protected abstract fun deserializeSingleMetric(value: Any?): ScalarType?
+
+    /**
+     * Implementor's provided function to serialized 'user' lifetime
+     * data to [String].
+     *
+     * @param value loaded from the storage as [ScalarType]
+     *
+     * @return data as [String] or null if serialization failed
+     */
+    protected fun serializeSingleMetric(value: ScalarType): String? {
+        return value.toString()
+    }
+
+    /**
+     * Deserialize the metrics with a lifetime = User that are on disk.
+     * This will be called the first time a metric is used or before a snapshot is
+     * taken.
+     *
+     * @return A [SharedPreferences] reference that will be used to inititialize [userLifetimeStorage]
+     */
+    @Suppress("TooGenericExceptionCaught")
+    open fun deserializeUserLifetime(): SharedPreferences {
+        val prefs =
+            applicationContext.getSharedPreferences(this.javaClass.simpleName, Context.MODE_PRIVATE)
+
+        val metrics = try {
+            prefs.all.entries
+        } catch (e: NullPointerException) {
+            // If we fail to deserialize, we can log the problem but keep on going.
+            logger.error("Failed to deserialize metric with 'user' lifetime")
+            return prefs
+        }
+
+        for ((metricName, metricValue) in metrics) {
+            if (!metricName.contains('#')) {
+                continue
+            }
+
+            // Split the stored name in 2: we expect it to be in the format
+            // store#metric.name
+            val parts = metricName.split('#', limit = 2)
+            if (parts[0].length <= 1) {
+                continue
+            }
+
+            val storeData = dataStores[Lifetime.User.ordinal].getOrPut(parts[0]) { mutableMapOf() }
+            // Only set the stored value if we're able to deserialize the persisted data.
+            deserializeSingleMetric(metricValue)?.let {
```

Comment:
If `deserializeSingleMetric()` returns `null`, are we logging an error somewhere?